### ISSUE TITLE
"Skipping tactic 0x0000000000000000 due to Myelin error" degrade performance.

### ISSUE THREAD

**Issue by YouSenRong** [NVIDIA/TensorRT#2838]
## Description

I want to run a model in FP16, but there is a FP16 subnormal value in a Constant layer "phase0_tf/predict_node/y:0", as follows:
<img width="941" alt="image" src="https://user-images.githubusercontent.com/21186538/229039591-113e1a59-275a-4448-9afd-44b49b823cac.png">

So I set this layer ("phase0_tf/predict_node/y:0") to FP32 by setPrecision(). Then I get the following message:

[V] [TRT] ConstShuffleFusion: Fusing phase0_tf/predict_node/y:0 with (Unnamed Layer* 565) [Shuffle]
......
[W] [TRT] Weights [name=phase0_tf/predict_node/y:0 + (Unnamed Layer* 565) [Shuffle]] had the following issues when converted to FP16:
[W] [TRT]  - Subnormal FP16 values detected. 
[W] [TRT] If this is not the desired behavior, please modify the weights or retrain with regularization to reduce the magnitude of the weights.
[W] [TRT] Skipping tactic 0x0000000000000000 due to Myelin error: Mismatched type for tensor (Unnamed Layer\_ 565) [Shuffle]_output', f32 vs. expected type:f16.
[V] [TRT] Fastest Tactic: 0xd15ea5edd15ea5ed Time: inf

By comparing setting or no setting this layer to FP32 precision, the performance does degrade, as follows:
||FP16|FP16+FP32|
|---|---|---|
|run time(ns)|4551286|5426231|

I am confusing of the relation of "(Unnamed Layer* 565) [Shuffle]" and "(Unnamed Layer\_ 565) [Shuffle]_output". What are differences between "\*" and "\_" in "Unnamed Layer"? The "phase0_tf/predict_node/y:0" layer is Fused with (Unnamed Layer* 565) [Shuffle], but why there is a "(Unnamed Layer\_ 565) [Shuffle]_output"?

It seems that the inserted Unnamed Layer "(Unnamed Layer* 565) [Shuffle]" makes something different, and the ConstShuffleFusion makes the setting "phase0_tf/predict_node/y:0" to FP32 precision failed or something else wrong.
As after I set "phase0_tf/predict_node/y:0" to FP32 precision, I also get the warning as follows:

[W] [TRT] Weights [name=phase0_tf/predict_node/y:0 + (Unnamed Layer* 565) [Shuffle]] had the following issues when converted to FP16:
[W] [TRT]  - Subnormal FP16 values detected. 

Besides, I have also try to set layer_precision and layer_output_type of "phase0_tf/predict_node/y:0" to FP32. It also doesn't make sense.

Thanks for any advices!

## Environment

**TensorRT Version**: 8.4.3
**NVIDIA GPU**: T4
**NVIDIA Driver Version**: 460.91.03
**CUDA Version**: 11.6
**CUDNN Version**: 
**Operating System**: 
**Python Version (if applicable)**: 
**Tensorflow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 
**Baremetal or Container (if so, version)**:

**Comment by zerollzeng** [comment]
> I am confusing of the relation of "(Unnamed Layer* 565) [Shuffle]" and "(Unnamed Layer_ 565) [Shuffle]_output". What are differences between "*" and "_" in "Unnamed Layer"? The "phase0_tf/predict_node/y:0" layer is Fused with (Unnamed Layer* 565) [Shuffle], but why there is a "(Unnamed Layer_ 565) [Shuffle]_output"?

Unnamed Layers is the layer generated by TensorRT internally. let's take an example, suppose you have a network like input->operator_1->output. if TRT decides to let it run in FP16, and the input is FP32, then TRT will insert a reformat layer to do the FP32->FP16 conversion automatically.

> [W] [TRT] - Subnormal FP16 values detected.
> [W] [TRT] If this is not the desired behavior, please modify the weights or retrain with regularization to reduce the magnitude of the weights.

This means the value can not be represented in FP16 precision, e.g. too small or too big. this might cause accuracy loss.

**Comment by YouSenRong** [comment]
Thanks for your reply!

> if TRT decides to let it run in FP16, and the input is FP32, then TRT will insert a reformat layer to do the FP32->FP16 conversion automatically

Yes, I know what the Unnamed layer is. What I am confused about is "(Unnamed Layer_ 565) [Shuffle]_output", I can only see there are "Unnamed Layer* ..." (such as "(Unnamed Layer* 565) [Shuffle]") during the network parsing, but the warning message contains "(Unnamed Layer_ 565) [Shuffle]", which contains "_", rather than "*". I am confused of the relation between "(Unnamed Layer* 565) [Shuffle]" and "(Unnamed Layer_ 565) [Shuffle]". 

> This means the value can not be represented in FP16 precision, e.g. too small or too big. this might cause accuracy loss.

I knows it, but I set this layer to compute in FP32 by set precision, there is warning:
[W] [TRT] Skipping tactic 0x0000000000000000 due to Myelin error: Mismatched type for tensor (Unnamed Layer_ 565) [Shuffle]_output', f32 vs. expected type:f16.
[V] [TRT] Fastest Tactic: 0xd15ea5edd15ea5ed Time: inf

This degrade the inference performance. Is there any approach to avoid this degradation? Thanks a lot!

**Comment by zerollzeng** [comment]
> This degrade the inference performance. Is there any approach to avoid this degradation?

I don't think there is WAR for this, it's rejected internally. Maybe we will improve this in the future version.

**Comment by YouSenRong** [comment]
Thanks a lot. And what about the relation of "(Unnamed Layer\* 565) [Shuffle]" and "(Unnamed Layer\_ 565) [Shuffle]"? Is there any relation?

> Yes, I know what the Unnamed layer is. What I am confused about is "(Unnamed Layer\_ 565) [Shuffle]_output", I can only see there are "Unnamed Layer\* ..." (such as "(Unnamed Layer\* 565) [Shuffle]") during the network parsing, but the warning message contains "(Unnamed Layer\_ 565) [Shuffle]", which contains "\_", rather than "\*". I am confused of the relation between "(Unnamed Layer\* 565) [Shuffle]" and "(Unnamed Layer\_ 565) [Shuffle]".

**Comment by pranavm-nvidia** [comment]
`(Unnamed Layer_ 565) [Shuffle]_output` is the output tensor of that layer.

**Comment by YouSenRong** [comment]
> `(Unnamed Layer_ 565) [Shuffle]_output` is the output tensor of that layer.

Thanks! So the "(Unnamed Layer_ 565) [Shuffle]_output" is the output tensor of the ConstantShuffleFusion node "[name=phase0_tf/predict_node/y:0 + (Unnamed Layer* 565) [Shuffle]]"? @pranavm-nvidia 

But I have set the layer "phase0_tf/predict_node/y:0" to FP32 by setPrecision(), the set build flags as OBEY. Why it also select the FP16 precision for the ConstantShuffleFusion node "phase0_tf/predict_node/y:0 + (Unnamed Layer* 565) [Shuffle]".

**Comment by pranavm-nvidia** [comment]
I think you probably want to set the operation's (i.e. the `Max` layer) precision to FP32 as well as probably the `Shuffle` preceding it. Just setting the precision of the constant itself doesn't constrain its consumers from using a different compute precision.

**Comment by YouSenRong** [comment]
> I think you probably want to set the operation's (i.e. the `Max` layer) precision to FP32 as well as probably the `Shuffle` preceding it.

Yes, as the constant is subnormal. 

> Just setting the precision of the constant itself doesn't constrain its consumers from using a different compute precision.

It consumers here is the Max or the probably Shuffle? I have also try to set the Max to FP32, but there is also warning:

> [W] [TRT] Skipping tactic 0x0000000000000000 due to Myelin error: Mismatched type for tensor (Unnamed Layer_ 565) [Shuffle]_output', f32 vs. expected type:f16.

Thanks.

**Comment by pranavm-nvidia** [comment]
You'd probably need to run both in FP32. Is the other input to the `Max` node still FP16? I'm wondering if that may be the reason for the warning.

**Comment by YouSenRong** [comment]
> You'd probably need to run both in FP32.

Yes, I run both the Constant Input "phase0_tf/predict_node/y:0" and the Max Layer both in FP32.

> Is the other input to the `Max` node still FP16?

But the other input to Max still in FP16. Do I need to set the other input to FP16? Isn't there TensorRT insert any "Unnamed Layer" to do this cast automatically?

> Unnamed Layers is the layer generated by TensorRT internally. let's take an example, suppose you have a network like input->operator_1->output. if TRT decides to let it run in FP16, and the input is FP32, then TRT will insert a reformat layer to do the FP32->FP16 conversion automatically.

**Comment by pranavm-nvidia** [comment]
Yeah, it should but seems like it's not in this case. You could either manually set the precision of the other input too or try a newer version of TensorRT since this may already be fixed.

**Comment by ttyio** [comment]
Closing since no activity for more than 3 weeks, pls reopen if you still have question, thanks!
